git/readtree.go

```diff
@@ -92,43 +100,183 @@ func ReadTreeMerge(c *Client, opt ReadTreeOptions, stage1, stage2, stage3 Treeis
+		ac := ours.Contains(path)
+		bc := theirs.Contains(path)
+
+		if oc && ac && bc {
+			oa := samePath(base, ours, path)
+			ob := samePath(base, theirs, path)
+
+			// t1000-read-tree-m-3way.sh test 75 "must match A in O && A && B && O!=A && O==B case.
+			// (This means we can't error out if the Sha1s dont match.)
+			if !oa && ob {
+				continue
+			}
+			if oa && !ob {
+				// Relevent cases:
+				// Must match and be up-to-date in O && A && B && O==A && O!=B
+				// May  match B in                 O && A && B && O==A && O!=B
+				b, ok := theirs[path]
+				if ok && b.Sha1 == orig.Sha1 {
+					continue
+				} else if !path.IsClean(c, o.Sha1) {
+					return idx, fmt.Errorf("Entry '%v' would be overwritten by a merge. Cannot merge.", path)
+				}
+			}
+		}
+		// Must match and be up-to-date in !O && A && B && A != B case test from AND
+		// Must match and be up-to-date in O && A && B && A != B case test from
+		// t1000-read-tree-m-3way.sh in official git
+		if ac && bc && !samePath(ours, theirs, path) {
+			if !path.IsClean(c, o.Sha1) {
+				return idx, fmt.Errorf("Entry '%v' would be overwritten by a merge. Cannot merge.", path)
+			}
+		}
+
+		// Must match and be up-to-date in O && A && !B && !B && O != A case AND
+		// Must match and be up-to-date in O && A && !B && !B && O == A case from
+		// t1000-read-tree-m-3way.sh in official git
+		if oc && ac && !bc {
+			if !path.IsClean(c, o.Sha1) {
+				return idx, fmt.Errorf("Entry '%v' would be overwritten by a merge. Cannot merge.", path)
+			}
+		}
+
+		if o.Sha1 != orig.Sha1 {
+			return idx, fmt.Errorf("Entry '%v' would be overwritten by a merge. Cannot merge.", path)
+		}
+	}
+	idx = NewIndex()
+paths:
+	for _, path := range allObjects {
+		// Handle directory/file conflicts.
+		if base.HasDir(path) || ours.HasDir(path) || theirs.HasDir(path) {
+			// Keep track of what was a directory so that other
+			// other paths know if they had a conflict higher
+			// up in the tree.
+			dirs = append(dirs, path)
+
+			// Add the non-directory version fo the appropriate stage
+			if p, ok := base[path]; ok {
+				idx.AddStage(c, path, p.Mode, p.Sha1, Stage1, p.Fsize, time.Now().UnixNano(), UpdateIndexOptions{Add: true})
+			}
+			if p, ok := ours[path]; ok {
+				idx.AddStage(c, path, p.Mode, p.Sha1, Stage2, p.Fsize, time.Now().UnixNano(), UpdateIndexOptions{Add: true})
+			}
+			if p, ok := theirs[path]; ok {
+				idx.AddStage(c, path, p.Mode, p.Sha1, Stage3, p.Fsize, time.Now().UnixNano(), UpdateIndexOptions{Add: true})
+			}
+			continue
+		}
+
+		// Handle the subfiles in any directory that had a conflict
+		// by just adding them in the appropriate stage, because
+		// there's no way for a directory and file to not be in
+		// conflict.
+		for _, d := range dirs {
+			if strings.HasPrefix(string(path), string(d+"/")) {
+				if p, ok := base[path]; ok {
+					if err := idx.AddStage(c, path, p.Mode, p.Sha1, Stage1, p.Fsize, time.Now().UnixNano(), UpdateIndexOptions{Add: true}); err != nil {
```

Comment:
Should this panic or should it return the error to the caller?

Reply:
I couldn't decide that when I was writing it. We *could* propagate the error, but  If read-tree can't add a stage to the index, something has gone horribly, horribly wrong and it probably needs investigating.

Reply:
I changed it to return an error, I'll merge it once it passes Travis.